#!/usr/bin/perl
use Getopt::Long;
use warnings;
#use strict;
GetOptions (\%opt,"gff:s","code:s", "anno:s", "project:s","help");


my $help=<<USAGE;
Assign gene code to gene and transcripts according to chromosome postion (read from code table generated by AssignGeneCode4chr.pl).
We start from ObXXg10010 for each chromosome and increace by 10 every gene.
The transcript id is ObXXg10010.1 for each gene since we have only one transcript for each gene.
perl $0 --gff --code --anno
--gff
chr11   maker   gene    6337489 6342862 .       +       .       ID=maker-scaffold_732-augustus-gene-0.271;Name=maker-scaffold_732-augustus-gene-0.271
chr11   maker   mRNA    6337489 6342862 .       +       .       ID=maker-scaffold_732-augustus-gene-0.271-mRNA-1;Parent=maker-scaffold_732-augustus-gene-0.271;Name=maker-scaffold_732-augustus-gene-0.271-mRNA-1;_AED=0.05;_eAED=0.05
chr11   maker   exon    6337489 6338288 .       +       .       ID=maker-scaffold_732-augustus-gene-0.271-mRNA-1:exon:971;Parent=maker-scaffold_732-augustus-gene-0.271-mRNA-1
chr11   maker   exon    6340894 6342862 .       +       .       ID=maker-scaffold_732-augustus-gene-0.271-mRNA-1:exon:972;Parent=maker-scaffold_732-augustus-gene-0.271-mRNA-1
chr11   maker   five_prime_UTR  6337489 6337500 .       +       .       ID=maker-scaffold_732-augustus-gene-0.271-mRNA-1:five_prime_utr;Parent=maker-scaffold_732-augustus-gene-0.271-mRNA-1
chr11   maker   CDS     6337501 6338288 .       +       0       ID=maker-scaffold_732-augustus-gene-0.271-mRNA-1:cds;Parent=maker-scaffold_732-augustus-gene-0.271-mRNA-1
chr11   maker   CDS     6340894 6342862 .       +       1       ID=maker-scaffold_732-augustus-gene-0.271-mRNA-1:cds;Parent=maker-scaffold_732-augustus-gene-0.271-mRNA-1

--code
maker-scaffold_540-snap-gene-0.463      HEG4_Os01g10010
maker-scaffold_540-snap-gene-0.463-mRNA-1       HEG4_Os01g10010.1

USAGE


if ($opt{help} or keys %opt < 1){
    print "$help\n";
    exit();
}

my $filtergff=$1 if ($opt{gff}=~/(.*)\.gff/);
my $code=readtable($opt{code});
my $anno=readanno($opt{anno});

open OUT, ">$filtergff.v1.gff" or die "$!";
open IN, "$opt{gff}" or die "$!";
while(<IN>){
    chomp $_;
    next if ($_=~/^#/);
    my @unit=split("\t",$_);
    if ($unit[2] eq 'gene'){
       if ($unit[8]=~/ID=(.*?);/){
          $unit[8]="ID=$code->{$1};Name=$anno->{$1};";
       }
    }elsif($unit[2] eq 'mRNA'){
       if ($unit[8]=~/ID=(.*?);Parent=(.*?);/){
          $unit[8]="ID=$code->{$1};Parent=$code->{$2};";
       }
    }else{
       if ($unit[8]=~/ID=(.*?):.*?;Parent=(.*?)/){
          $unit[8]="ID=$code->{$1}:$unit[2];Parent=$code->{$1};";
       }
    }
    my $geneline = join("\t", @unit);
    print OUT "$geneline\n";
}
close IN;
close OUT;


sub readanno
{
my ($file)=@_;
my %hash;
open IN, "$file" or die "$!";
while(<IN>){
    chomp $_;
    next if ($_=~/^$/);
    my @unit=split("\t",$_);
    #print "$unit[0]\n";
    my $gene = $1 if $unit[0]=~/(.*)-mRNA-\d+/;
    $hash{$gene}=$unit[1];
}
close IN;
return \%hash;
}



sub readtable
{
my ($file)=@_;
my %hash;
open IN, "$file" or die "$!";
while(<IN>){
    chomp $_;
    next if ($_=~/^$/);
    my @unit=split("\t",$_);
    $hash{$unit[0]}=$unit[1];
}
close IN;
return \%hash;
}

